### ISSUE TITLE
edge_cylinders: something wrong while running readNODE 

### ISSUE THREAD

**Issue by Frank-Dz** [alecjacobson/gptoolbox#89]
Thanks to your help, my last problem solved > https://github.com/alecjacobson/gptoolbox/issues/88
And now I successfully installed triangle.exe and it did work.
Yet, when I went next step, I met this problem:

ERROR: readNODE (line 23)
  header = fscanf(node_file_handle,'%d %d %d %d',4);

ERROR: triangle (line 231)
  [TV,I] = readNODE([prefix '.1.node']);

ERROR: fill_holes (line 28)
    [EVV,EF] = triangle(EV*A,E,[],'Quiet');

ERROR: edge_cylinders (line 220)
  CF = [CF;fill_holes(CV,CF)];

Again my obj file is as follows:
```
v 0 0.850651 0.525731
v 0 0.680521 0.420585v 0 1.02078 0.630877
l 1 2
v 0 0.850651 -0.525731
v 0 0.680521 -0.420585v 0 1.02078 -0.630877
l 3 4
v 0 -0.850651 0.525731
v 0 -0.680521 0.420585v 0 -1.02078 0.630877
l 5 6
v 0 -0.850651 -0.525731
v 0 -0.680521 -0.420585v 0 -1.02078 -0.630877
l 7 8
v 0.850651 0.525731 0
v 0.680521 0.420585 0v 1.02078 0.630877 0
l 9 10
v 0.850651 -0.525731 0
v 0.680521 -0.420585 0v 1.02078 -0.630877 0
l 11 12
v -0.850651 0.525731 0
v -0.680521 0.420585 0v -1.02078 0.630877 0
l 13 14
v -0.850651 -0.525731 0
v -0.680521 -0.420585 0v -1.02078 -0.630877 0
l 15 16
v 0.525731 0 0.850651
v 0.420585 0 0.680521v 0.630877 0 1.02078
l 17 18
v -0.525731 0 0.850651
v -0.420585 0 0.680521v -0.630877 0 1.02078
l 19 20
v 0.525731 0 -0.850651
v 0.420585 0 -0.680521v 0.630877 0 -1.02078
l 21 22
v -0.525731 0 -0.850651
v -0.420585 0 -0.680521v -0.630877 0 -1.02078
l 23 24
```
Is it due to the wrong .OBJ format? Or is there a demo to demonstrate how to use this 'edge_cylinders' function?

Best,
Frank

**Comment by Frank-Dz** [comment]
The input obj file:
![snapshot00](https://user-images.githubusercontent.com/55009565/72071007-cf1fdd80-3325-11ea-82d9-975d2d2d03fb.png)

**Comment by bbrrck** [comment]
I think your .obj file is fine. There is some problem with reading the .node file output by Triangle. Try removing the 'Quiet' flag when Triangle is called (line 28 of fill_holes.m according to your log) to see if the generated mesh is not empty. If the mesh is fine, the problem might be that the .node file is written into a wrong location or not at all.

**Comment by Frank-Dz** [comment]
> I think your .obj file is fine. There is some problem with reading the .node file output by Triangle. Try removing the 'Quiet' flag when Triangle is called (line 28 of fill_holes.m according to your log) to see if the generated mesh is not empty. If the mesh is fine, the problem might be that the .node file is written into a wrong location or not at all.

Hi~ Thanks for you kind guidance. Yep. I did try to locate the file:
![image](https://user-images.githubusercontent.com/55009565/72071740-75b8ae00-3327-11ea-8edd-e956f63b455f.png)
but
![image](https://user-images.githubusercontent.com/55009565/72071860-b1537800-3327-11ea-8a33-bd4b432f85a4.png)

**Comment by Frank-Dz** [comment]
![image](https://user-images.githubusercontent.com/55009565/72071941-e069e980-3327-11ea-81c2-88167733c28f.png)
Tried. It doesn't work.

**Comment by Frank-Dz** [comment]
I think I figure it out:
![image](https://user-images.githubusercontent.com/55009565/72072069-28890c00-3328-11ea-9ac6-5cdd7b077f0d.png)
This is due to that there is a '.1.' added into the path?

**Comment by bbrrck** [comment]
My bad, the output is disabled by default, you have to do `[EVV,EF] = triangle(EV*A,E,[],'Quiet',0);`

Are you using `readOBJ.m` from this repo? That one is not reading the edges, so you might be passing empty input to `edge_cylinders.m`.

**Comment by Frank-Dz** [comment]
> My bad, the output is disabled by default, you have to do `[EVV,EF] = triangle(EV*A,E,[],'Quiet',0);`
> 
> Are you using `readOBJ.m` from this repo? That one is not reading the edges, so you might be passing empty input to `edge_cylinders.m`.

I learnt that readOBJ is not able to handle edges, Thus I write a function which can process the situation. 
![image](https://user-images.githubusercontent.com/55009565/72072805-ac8fc380-3329-11ea-9c70-b6c8a996bde9.png)

Plus,
![image](https://user-images.githubusercontent.com/55009565/72072856-bc0f0c80-3329-11ea-9ec0-ffd8b28c74a5.png)
with '0' added, it still doesn't work. : (

**Comment by bbrrck** [comment]
hm that's weird, maybe try `false` instead of `0`?

**Comment by Frank-Dz** [comment]
> hm that's weird, maybe try `false` instead of `0`?
Still doesn't make it. Is there an alternative for this process (edge2cylinder) ?
![Uploading image.png…]()

**Comment by bbrrck** [comment]
What's the error?

**Comment by bbrrck** [comment]
Also, have you tried a minimal example? Something like this:
```
PV = [1 0 0; 0 1 0];
PE = [1 2];
[CV,CF,CJ,CI] = edge_cylinders(PV,PE);
```

**Comment by Frank-Dz** [comment]
> What's the error?

Still, The code is trying to find 
C:\Users\Frank\AppData\Local\Temp\tpae6bc758_90a8_44f8_984d_85befae04e08.ele

but This .ele file doesn't exist.

**Comment by Frank-Dz** [comment]
Tried:
```
 PV = [1 0 0; 0 1 0];
PE = [1 2];
[CV,CF,CJ,CI] = edge_cylinders(PV,PE);
```
Same error：
The code is trying to find
C:\Users\Frank\AppData\Local\Temp\tp873851b3_88af_4f0d_b169_f31df467d16a.ele
but this .ele file doesn't exist.

**Comment by bbrrck** [comment]
In that case it seems like the mesh given by Triangle is empty.

What is the error when you set the 'Quiet' option to false? If that still doesn't work (which is weird), you can bypass it by setting `quiet = false;` directly in `triangle.m`.

**Comment by Frank-Dz** [comment]
> In that case it seems like the mesh given by Triangle is empty.
> 
> What is the error when you set the 'Quiet' option to false? If that still doesn't work (which is weird), you can bypass it by setting `quiet = false;` directly in `triangle.m`.

I tried set quiet  to be false, and just couldn't generate .ele files :
C:\Users\Frank\AppData\Local\Temp\tp47b8602f_4746_4ff7_9a01_9ed41bcc41be.ele

The error is that couldn't open this file:
![image](https://user-images.githubusercontent.com/55009565/72074949-c92dfa80-332d-11ea-831b-1f6060ff7adf.png)

**Comment by bbrrck** [comment]
Can you post the output given by Triangle when you set `quiet = false`? 
It should be something like this:

```
Opening /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.poly.
Opening /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.node.
Constructing Delaunay triangulation by divide-and-conquer method.
Delaunay milliseconds:  0
Recovering segments in Delaunay triangulation.
Segment milliseconds:  0
Removing unwanted triangles.
Hole milliseconds:  0

Writing /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.1.node.
Writing /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.1.ele.
Writing /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.1.poly.

Output milliseconds:  0
Total running milliseconds:  1

Statistics:

Input vertices: 4
Input segments: 4
Input holes: 0

Mesh vertices: 4
Mesh triangles: 2
Mesh edges: 5
Mesh exterior boundary edges: 4
Mesh interior boundary edges: 0
Mesh subsegments (constrained edges): 4
```

**Comment by Frank-Dz** [comment]
> Can you post the output given by Triangle when you set `quiet = false`?
> It should be something like this:
> 
> ```
> Opening /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.poly.
> Opening /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.node.
> Constructing Delaunay triangulation by divide-and-conquer method.
> Delaunay milliseconds:  0
> Recovering segments in Delaunay triangulation.
> Segment milliseconds:  0
> Removing unwanted triangles.
> Hole milliseconds:  0
> 
> Writing /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.1.node.
> Writing /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.1.ele.
> Writing /private/tmp/tp2326acc6_0596_41bc_806d_f13c825fb369.1.poly.
> 
> Output milliseconds:  0
> Total running milliseconds:  1
> 
> Statistics:
> 
> Input vertices: 4
> Input segments: 4
> Input holes: 0
> 
> Mesh vertices: 4
> Mesh triangles: 2
> Mesh edges: 5
> Mesh exterior boundary edges: 4
> Mesh interior boundary edges: 0
> Mesh subsegments (constrained edges): 4
> ```

I just copy the command generated by MATLAB:
![image](https://user-images.githubusercontent.com/55009565/72075590-0b0b7080-332f-11ea-9ed5-337c75163df4.png)

which is :
```
C:/local/Triangle/triangle.exe -p  C:\Users\Frank\AppData\Local\Temp\tp99086336_d1b8_4621_ad67_f50989cfd147
```

And I run it in Cygwin64:
![image](https://user-images.githubusercontent.com/55009565/72075560-f9c26400-332e-11ea-94d2-fba46a00d0a5.png)

**Comment by bbrrck** [comment]
I don't understand, why do you run it from command line?

The output I'm looking for is stored in the `result` variable in the screenshot you sent.

**Comment by Frank-Dz** [comment]
> I don't understand, why do you run it from command line?
> 
> The output I'm looking for is stored in the `result` variable in the screenshot you sent.

Actually, I am a little confused. Should I just run
```
[EVV,EF] = triangle(EV*A,E,[],'Quiet',false);
```

**Comment by bbrrck** [comment]
Yes. If you set the `'Quiet'` flag to `false`, you'll get a full Triangle output directly in Matlab. Since it didn't work for you for some reason, you can either modify the flag manually in the `triangle.m` file, or just do `disp(result)` after Triangle is called, i.e. after the line `[status, result] = system( command );` in `triangle.m`.

**Comment by Frank-Dz** [comment]
> Yes. If you set the `'Quiet'` flag to `false`, you'll get a full Triangle output directly in Matlab. Since it didn't work for you for some reason, you can either modify the flag manually in the `triangle.m` file, or just do `disp(result)` after Triangle is called, i.e. after the line `[status, result] = system( command );` in `triangle.m`.

I am not sure that am I right ? 
![image](https://user-images.githubusercontent.com/55009565/72079613-380f5180-3336-11ea-91db-5941b5cc4aec.png)


![image](https://user-images.githubusercontent.com/55009565/72079673-4eb5a880-3336-11ea-8275-0dce23de4e54.png)

That means the output ‘result’ is null.

**Comment by bbrrck** [comment]
That is right, the output is empty, which means the call to Triangle is not executed properly. I suggest you have a look at the [instructions for Triangle](https://www.cs.cmu.edu/~quake/triangle.html) to first make it work outside of Matlab on a simple example, such as a .node file with three input points.

**Comment by Frank-Dz** [comment]
ok I will have a try. Plus, have you tried running the prog on you PC？ Is the result of the prog ok on your machine? Thanks again for you kind help!

**Comment by Desideriuswang** [comment]
thank you




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Zhiyang-Dou"<notifications@github.com&gt;;
发送时间:&nbsp;2020年1月13日(星期一) 下午4:14
收件人:&nbsp;"alecjacobson/gptoolbox"<gptoolbox@noreply.github.com&gt;;
抄送:&nbsp;"Subscribed"<subscribed@noreply.github.com&gt;;
主题:&nbsp;Re: [alecjacobson/gptoolbox] edge_cylinders: something wrong while running readNODE  (#89)




Closed #89.
